src/finch/compile/bufferized_ndarray.py

```diff
@@ -7,46 +9,99 @@
 from .. import finch_notation as ntn
 from ..algebra import Tensor
 from ..codegen import NumpyBuffer
-from ..finch_assembly import AssemblyStructFType
+from ..finch_assembly import AssemblyStructFType, TupleFType, Variable
 from ..symbolic import ftype
 from . import looplets as lplt
 from .lower import FinchTensorFType
 
 
+class StrideState(Enum):
+    FREE = 0
+    ACTIVE = 1
+    VISITED = 2
+
+
+class BufferizedNDArrayStrideState:
```

Comment:
The rationale behind StrideState is that we need to support transpositions/dims permutations.

We need to track which indices we visited to know which strides to use in each loop. A good example is:
```
for i = _
    for j = _
        A(i, j)
        B(j, i) 
```
when visiting loops we first have `[i, j]` and then descending to leaves `[j]` (or the other way around). Then when we access strides it's a different stride for A and for B. But if `self.nind: int` is all we have then we're going to choose the same stride idxs for both which is incorrect.

`BufferizedNDArrayStrideState` allow us to track which strides we visited and which are active in a given descent level.